### ISSUE TITLE
Set Current Version and Compatibility Version

### ISSUE THREAD

**Issue by mrackwitz** [realm/realm-cocoa#2180]
We should ensure that we set the current versions in the MachO headers of the binaries of both of our frameworks. For `RealmSwift.framework`, we should also set the version of `Realm.framework` we depend upon, which is very likely always the same version.

#### Problem

In that way we can make sure that when our users integrate different versions, e.g. because they updated only one of the prebuilt binaries from the release package, that the build will either fail preemptively on link-time or at least on dynamic link-time when the executable is loaded by `dyld`. Otherwise the build could succeed and instead the app may crash in a cryptic way on runtime.

#### Current State

That's what we currently have:
```shell
> otool -L RealmSwift.framework/RealmSwift | grep Realm
RealmSwift.framework/RealmSwift:
	@rpath/RealmSwift.framework/RealmSwift (compatibility version 1.0.0, current version 1.0.0)
	@rpath/RealmSwift.framework/Frameworks/Realm.framework/Realm (compatibility version 1.0.0, current version 1.0.0)
```

#### Versioning System

We would have to think about a way of mapping our release versions, which conform to [SemVer](https://semver.org), to the in comparison more restricted versioning system available for dynamic linking. This builds upon the following components:
* a major revision, usually identified by the letter `A`, which no one outside Apple and not even them consequently uses. This is implemented on the file system base via different named / placed executables in the bundle and only available on OS X. In that way it is also included in the MachO, but the dynamic linker doesn't have any further knowledge of those versions.
* a minor version / compatibility version `X[.Y[.Z]]`, where `X` is a number between 0 and 65535, and `Y` and `Z` are numbers between 0 and 255.

Because major revisions means retaining backwards-compatibility, which is not practicable for distribution, they don't really qualify to map on, so we should align to keeping only one major revision `A`. That means, we can only consider the minor version, which can mean breaking changes as well. The minor version can only be mapped to a true subset of SemVer, because the latter allows to express more with prereleases and build numbers. As we don't use them yet, we should be fine.

See [Dynamic Library Programming Topics > Dynamic Library Design Guidelines  > Managing Client Compatibility With Dependent Libraries > Specifying Version Information](https://developer.apple.com/library/mac/documentation/DeveloperTools/Conceptual/DynamicLibraries/100-Articles/DynamicLibraryDesignGuidelines.html#//apple_ref/doc/uid/TP40002013-SW23).

##### Build Settings

The versions can be influenced by `xcodebuild` via the following build settings:
* `DYLIB_COMPATIBILITY_VERSION`
* `DYLIB_CURRENT_VERSION` which defaults to `$(CURRENT_PROJECT_VERSION)` which holds the actual value and is set by default to `1`.

##### Real Life Usage

Apple aligns to its versioning system with the Objective-C stdlib:
```shell
> otool -L RealmSwift.framework/RealmSwift | grep Foundation.framework
	/System/Library/Frameworks/CoreFoundation.framework/CoreFoundation (compatibility version 150.0.0, current version 1144.17.0)
	/System/Library/Frameworks/Foundation.framework/Foundation (compatibility version 300.0.0, current version 1144.17.0)
```

Interesting is the approach they took for the Swift stdlib versions:
```shell
> otool -L RealmSwift.framework/RealmSwift | grep swift
	@rpath/libswiftCore.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libswiftDarwin.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libswiftDispatch.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libswiftFoundation.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libswiftObjectiveC.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libswiftSecurity.dylib (compatibility version 0.0.0, current version 0.0.0)
```

**Comment by bdash** [comment]
A couple of observations:

* Framework versioning (e.g., `Versions/A` in the framework bundle) is not available on iOS, and is effectively unused and unsupported on OS X. It was originally intended to allow multiple ABI-incompatible versions of a framework to co-exist on a system, with applications loading the version that they were linked against. That plan never really played out on OS X, though a similar approach is used in rare cases for 3rd-party dylibs that Apple bundles with OS X. For instance, `libssl.dylib` is a symlink to `libssl.0.9.8.dylib`, and `libssl.0.9.7.dylib` is also shipped.

  Given that the primary purpose of framework versioning was to allow multiple library versions to co-exist within a single framework, it's not something that's relevant to how Realm is deployed in applications.

* The dylib current version is informational only. Only the compatibility version is checked by the dynamic loader.

* dylib versioning provides only a lower bound on the version of a dependent dylib that can be loaded. That is, a copy of `RealmSwift.framework` built against a  `Realm.framework` that specifies a compatibility version of `0.99.0` can load a version of `Realm.framework` with a compatibility version of `0.93.0`.

  I believe that this makes dylib versioning on its own insufficient to address the stated problem.

* Apple's use of dylib versioning in its own frameworks and libraries doesn't follow the guidelines set out in [Managing Client Compatibility With Dependent Libraries](https://developer.apple.com/library/mac/documentation/DeveloperTools/Conceptual/DynamicLibraries/100-Articles/DynamicLibraryDesignGuidelines.html). The need for applications built against SDK vN to run on OS vN-1 means that they cannot increase the compatibility version field when the ABI is changed (e.g., new functions are added), since that would lead to the dynamic loader rejecting vN-1's version of a library as being too old. You'll notice that all frameworks added to iOS and OS X in recent years have a compatibility version of 1, while older frameworks are fixed at their then-current version numbers. See [QA1659](https://developer.apple.com/library/ios/qa/qa1659/_index.html) for a case where we got this wrong.

  This issue clearly doesn't apply to Realm since we can safely assume that the version of Realm being used at runtime is the same version (or newer) than the version that was present at link time.

**Comment by mrackwitz** [comment]
Thanks for providing all those infos. :heart:


You wrote:
>The dylib current version is informational only. Only the compatibility version is checked by the dynamic loader.
>
>dylib versioning provides only a lower bound on the version of a dependent dylib that can be loaded. That is, a copy of RealmSwift.framework built against a Realm.framework that specifies a compatibility version of 0.99.0 can load a version of Realm.framework with a compatibility version of 0.93.0.

But I think, the example is wrong, isn't it? I think, it should be:
A copy of RealmSwift.framework built against a Realm.framework that specifies a compatibility version of **0.93.0** can load a version of Realm.framework with a compatibility version of **0.99.0**.

Because the docs clearly state:
>Before loading a dynamic library, the dynamic loader compares the current version of the .dylib file in the user’s file system with the compatibility version of the .dylib file the client was linked with in the developer’s file system. If the current version is earlier (less) than the compatibility version, the dependent library is not loaded. Therefore, the launch process (for client apps) or the load process (for client libraries) is aborted.

And dyld actually does something else, as you describe in the first two sentences:

```c++
// in ImageLoaderMachO::doGetDependentLibraries(DependentLibraryInfo libs[]):
lib->info.minVersion = dylib->dylib.compatibility_version;
// in ImageLoader::recursiveLoadLibraries(const LinkContext& context, bool preflightOnly, const RPathChain& loaderRPaths):
if ( (requiredLibInfo.info.minVersion != 0xFFFFFFFF) && (actualInfo.minVersion < requiredLibInfo.info.minVersion) ) {
    // record values for possible use by CrashReporter or Finder
   dyld::throwf("…")
```

In conclusion I wouldn't fully agree with that:

>I believe that this makes dylib versioning on its own insufficient to address the stated problem.

We can be at least be a little bit more helpful, if users update their `RealmSwift.framework`, but don't update their `Realm.framework`.

**Comment by bdash** [comment]
> But I think, the example is wrong, isn't it?

Yes, sorry. I reversed the order of the version numbers in my example :-(

> In conclusion I wouldn't fully agree with that:

I think we agree that it addresses only one side of the problem: a user updating `RealmSwift.framework` but not `Realm.framework`. It won't help a user that updates `Realm.framework` but not `RealmSwift.framework`. I believe addressing the latter would require us to perform explicit runtime version checks.

**Comment by segiddins** [comment]
Given the problems we've seen in CocoaPods with setting these version fields, I'd be against doing this.

**Comment by bdash** [comment]
@segiddins, can you provide a pointer to any information about the problems you're referring to?

**Comment by segiddins** [comment]
https://github.com/CocoaPods/CocoaPods/issues/3903, https://github.com/CocoaPods/CocoaPods/issues/3940, and I'm sure there are more. We're even talking about reverting the feature entirely in CocoaPods, as it provides very minimal benefit.

**Comment by jpsim** [comment]
CocoaPods stopped setting these in #4272, is this still something we want to do? @mrackwitz @bdash

**Comment by bdash** [comment]
I don't think there's any reason to do this.

**Comment by mrackwitz** [comment]
I agree, there seems to be no need to do this.
